### ISSUE TITLE
Failed to locate opened application window with appId

### ISSUE THREAD

**Issue by Gaddale** [microsoft/WinAppDriver#1406]
Hi All,
I know this is repeated question, I have gone through the same issues logged my others and followed the solutions but some how I am not able to fix it. Request for help

Logs:
-------------
POST /session HTTP/1.1
Accept-Encoding: gzip
Connection: Keep-Alive
Content-Length: 475
Content-Type: application/json; charset=utf-8
Host: 127.0.0.1:4723
User-Agent: selenium/4.0.0-alpha-1 (java windows)

{
  "desiredCapabilities": {
    "app": "C:\\Program Files (x86)\\XXX\\XXX.exe",
    "appArguments": "-internal -NoSplash",
    "platformName": "Windows",
    "deviceName": "Windows10"
  },
  "capabilities": {
    "firstMatch": [
      {
        "appium:app": "C:\\Program Files (x86)\\XXX\\XXX.exe",
        "appArguments": "-internal -NoSplash",
        "appium:deviceName": "Windows10",
        "platformName": "windows"
      }
    ]
  }
}
HTTP/1.1 500 Internal Error
Content-Length: 190
Content-Type: application/json

{"status":13,"value":{"error":"unknown error","message":"Failed to locate opened application window with appId: C:\\Program Files (x86)\\XXX\\XXX.exe, and processId: 11564"}}

Code:
------------
DesiredCapabilities cap = new DesiredCapabilities();
cap.setCapability("app", "C:\\Program Files (x86)\\XXX\\XXX.exe");
cap.setCapability("platformName", "Windows");
cap.setCapability("deviceName", "Windows10");
cap.setCapability("appArguments", "-internal -NoSplash");
driver = new WindowsDriver(new URL("http://127.0.0.1:4723"), cap);
Assert.assertNotNull(driver);
driver.manage().timeouts().implicitlyWait(100, TimeUnit.SECONDS);

The test is failing but the app is actually getting launched.
Is there something I am doing wrong? 

WindowsDriver (v1.2.1)

Regards,
Nagaraj

**Comment by thinski** [comment]
try `cap.setCapability("app", "C:\\Program Files (x86)\\XXX\\XXX.exe");`

**Comment by Gaddale** [comment]
Hi Thinski,
I have tried, but no luck..

**Comment by anunay1** [comment]
From the error it looks like your application has a splash screen

**Comment by Gaddale** [comment]
Yes it has splash screen, that is the reason I have added the appArguments as ` "appArguments": "-internal -NoSplash",`

**Comment by anunay1** [comment]
Which language binding are you using

**Comment by Gaddale** [comment]
I am using Java, I have not tried with root session. Generally I feel its good if I try to launch the application via driver instead of manually opening and switching from root session.

**Comment by Gaddale** [comment]
Hi All,
Any more suggestion's..no luck from myside

**Comment by anunay1** [comment]
Try with the below code:
`try {
			DesiredCapabilities appCapabilities = new DesiredCapabilities();
			appCapabilities.setCapability("app", "Your APP");
			appCapabilities.setCapability("platformName", "Windows");
			appCapabilities.setCapability("deviceName", "WindowsPC");
			session = new WindowsDriver<WindowsElement>(new URL(WindowsApplicationDriverUrl), appCapabilities);
			Thread.sleep(20000);
			session.findElementByName("Any element");// this will throw and exception
			
		} catch (Exception e) {

				if ("Failed to locate opened application window with appId"))
				{
					DesiredCapabilities capabilities = new DesiredCapabilities();
					capabilities.setCapability("app", "Root");
					desktopsession = new WindowsDriver<WindowsElement>(new
					URL(WindowsApplicationDriverUrl),capabilities);


					String topWindow = desktopsession.findElementByClassName("Your application Main WIndow classname").getAttribute("NativeWindowHandle");
					int MAWinHandleInt = Integer.parseInt(topWindow);
					String MAWinHandleHex = Integer.toHexString(MAWinHandleInt);

					DesiredCapabilities caps = new DesiredCapabilities();
					caps.setCapability("appTopLevelWindow", MAWinHandleHex);
					session = new WindowsDriver<WindowsElement>(new URL(WindowsApplicationDriverUrl), caps);

				}

		}`

**Comment by anunay1** [comment]
if the exception is different remove that if statement

**Comment by vagisha-nidhi195** [comment]
@[Gaddale](https://github.com/Gaddale) Did the above solve the issue?

**Comment by anunay1** [comment]
Or have you already tried my code.

**Comment by anunay1** [comment]
Or have you already tried my code snippet

**Comment by vagisha-nidhi195** [comment]
@anunay1 I have a UWP app which I am opening up using the same method. Whenever the remote machine is locked, I get this error. If I am on RDP with the remote machine (unloacked), this works fine. I tried getting the desktop session, and I do get my app in this session, but I do not get any `NativeWindowHandle` attribute in my page source. Is it not supposed to be in the page source?

**Comment by anunay1** [comment]
@vagisha-nidhi195 UI automation does not work when the system is locked. This holds good for any automation setup, disable the lock.

**Comment by anunay1** [comment]
Also you can try the [Solution](https://support.smartbear.com/testcomplete/docs/testing-with/running/via-rdp/in-minimized-window.html) if you want to run the automation in RDP minimized mode.

**Comment by vagisha-nidhi195** [comment]
@anunay1 What do you mean by disable the lock ? I set my settings to "Never Lock", "Never sleep", I am still getting the same issue. Is there anything else to be done?

When I RDP into my computer, I don't see a lock screen anymore. But the test is still not working when I do not have an active RDP session.

**Comment by shruthibhatjm** [comment]
> Try with the below code: ` try { DesiredCapabilities appCapabilities = new DesiredCapabilities(); appCapabilities.setCapability("app", "Your APP"); appCapabilities.setCapability("platformName", "Windows"); appCapabilities.setCapability("deviceName", "WindowsPC"); session = new WindowsDriver(new URL(WindowsApplicationDriverUrl), appCapabilities); Thread.sleep(20000); session.findElementByName("Any element");// this will throw and exception
> 
> ```
> 	} catch (Exception e) {
> 
> 			if ("Failed to locate opened application window with appId"))
> 			{
> 				DesiredCapabilities capabilities = new DesiredCapabilities();
> 				capabilities.setCapability("app", "Root");
> 				desktopsession = new WindowsDriver<WindowsElement>(new
> 				URL(WindowsApplicationDriverUrl),capabilities);
> 
> 
> 				String topWindow = desktopsession.findElementByClassName("Your application Main WIndow 
>                     classname").getAttribute("NativeWindowHandle");
> 				int MAWinHandleInt = Integer.parseInt(topWindow);
> 				String MAWinHandleHex = Integer.toHexString(MAWinHandleInt);
> 
> 				DesiredCapabilities caps = new DesiredCapabilities();
> 				caps.setCapability("appTopLevelWindow", MAWinHandleHex);
> 				session = new WindowsDriver<WindowsElement>(new URL(WindowsApplicationDriverUrl), caps);
> 
> 			}
> ```
> 
> } `

What should be put in the 8th line of given code (" Any element") ?

**Comment by anunay1** [comment]
Class name of your main window, you will get that in inspect.exe

**Comment by shruthibhatjm** [comment]
> Class name of your main window, you will get that in inspect.exe

 session.findElementByName("Any element");

I am talking about this line of code. Over here do we need to put class name of the main window or the splash screen?

**Comment by anunay1** [comment]
Any element of the application?

**Comment by shruthibhatjm** [comment]
can I use the name or class name of the splash screen?

**Comment by shruthibhatjm** [comment]
desktopsession = new WindowsDriver<WindowsElement>(new
				URL(WindowsApplicationDriverUrl),capabilities);


is it desktopsession of session? because you have used session in the other part

**Comment by anunay1** [comment]
> can I use the name or class name of the splash screen?
> 

No.

**Comment by anunay1** [comment]
> desktopsession = new WindowsDriver<WindowsElement>(new
> 				URL(WindowsApplicationDriverUrl),capabilities);
> 
> 
> is it desktopsession of session? because you have used session in the other part

It's a seperate session?

**Comment by shruthibhatjm** [comment]
desktopsession should be initialized first right?
 and ("Any Element") can be an element on any screen like login screen too?

**Comment by anunay1** [comment]
Which block try or catch?

**Comment by shruthibhatjm** [comment]
in try block, in this line of code
session.findElementByName("Any element");// this will throw and exception

**Comment by anunay1** [comment]
Where is desktop session in the try block ? You only have session. And you have to give any element of the opened application window which will throw an exception and it goes to the catch block?

**Comment by shruthibhatjm** [comment]
desktop session in the catch block

**Comment by anunay1** [comment]
> desktop session in the catch block 
> 

You are confused, understand the code first before putting the question can't explain the same thing time and again
